contracts/src/GaspToken.sol

```diff
@@ -1,30 +1,126 @@
+                revert OperationForbidden(selector);
+            }
+        }
+        _;
     }
 
-    function mint(address account, uint256 amount) public {
-      require((owner == msg.sender), "Only one who deployed contract can mint tokens");
-      _mint(account, amount);
+    constructor(address l1Council_) Ownable() ERC20(_NAME, _SYMBOL) {
+        if (l1Council_ == address(0)) {
+            revert ZeroL1Council();
+        }
+
+        _transferOwnership(l1Council_);
+        _mint(l1Council_, _TOTAL_SUPPLY);
+
+        whitelist[l1Council_] = true;
     }
-}
 
+    function setUniswapPool(address uniswapPool_) external override onlyOwner {
+        if (uniswapPool_ == address(0)) {
+            revert ZeroUniswapPool();
+        }
 
-contract Gasp is MintableERC20 {
-    constructor() MintableERC20("GaspV2", "GASPV2") {}
-}
+        uniswapPool = uniswapPool_;
+        whitelist[uniswapPool_] = true;
+        emit UniswapPoolSet(uniswapPool_);
+    }
 
-contract DevToken is MintableERC20 {
-    constructor() MintableERC20("GaspDev", "GASPDEV") {}
-}
+    function setAllowTransfers(bool allowTransfers_) external override onlyOwner {
+        if (allowTransfers) {
+            revert TransfersAlreadyAllowed();
+        }
+
+        allowTransfers = allowTransfers_;
+        emit AllowTransfersSet(allowTransfers_);
+    }
+
+    function addToWhitelist(address account) external override onlyOwner {
+        if (account == address(0)) {
+            revert ZeroWhitelistAccount();
+        }
+        if (whitelist[account]) {
+            revert AccountAlreadyWhitelisted(account);
+        }
+
+        whitelist[account] = true;
+        emit AddedToWhitelist(account);
+    }
 
-contract ArbToken is MintableERC20 {
-    constructor() MintableERC20("CrossChain Token", "CCXT") {}
-}
\ No newline at end of file
+    function removeFromWhitelist(address account) external override onlyOwner {
+        if (account == address(0)) {
+            revert ZeroWhitelistAccount();
+        }
+        if (!whitelist[account]) {
+            revert AccountNotWhitelisted(account);
+        }
+
+        delete whitelist[account];
+        emit RemovedFromWhitelist(account);
+    }
+
+    function approve(address spender, uint256 amount)
+        public
+        override(ERC20, IERC20)
+        isWhitelisted([_msgSender(), spender], IERC20.approve.selector)
+        returns (bool)
+    {
+        return super.approve(spender, amount);
+    }
+
+    function transfer(address recipient, uint256 amount)
+        public
+        override(ERC20, IERC20)
+        isWhitelisted([_msgSender(), recipient], IERC20.transfer.selector)
+        returns (bool)
+    {
+        return super.transfer(recipient, amount);
+    }
+
+    function transferFrom(address owner, address recipient, uint256 amount)
+        public
+        override(ERC20, IERC20)
+        isWhitelisted([_msgSender(), recipient], IERC20.transferFrom.selector)
```

Comment:
The current special treatment of the uniswap pool allows everyone (including non-whitelisted addresses) to sell the token. It is the opposite of what we want to achieve: only whitelisted addresses are allowed to sell, and anyone is allowed to buy.